dbuf-lsp/src/action_handler/mod.rs

```diff
@@ -0,0 +1,182 @@
+            return format_errors::bad_trim_final_newlines();
+        }
+
+        let mut edit = TextEdit {
+            range: Range::new(Position::new(0, 0), Position::new(2e9 as u32, 0)),
+            new_text: String::new(),
+        };
+
+        let file = access.read(document);
+        let ast = file.get_parsed();
+
+        let mut writer = PrettyPrinter::new(&mut edit.new_text).with_tab_size(options.tab_size as usize);
+        writer.print_ast(ast);
+
+        Ok(Some(vec![edit]))
+    }
+
+    /// `textDocument/prepareRename` implementation.
+    ///
+    /// Currently checks if symbol can be renamed and,
+    /// if so, caches it.
+    ///
+    pub fn prepare_rename(
+        &self,
+        access: &WorkspaceAccess,
+        pos: Position,
+        document: &Url,
+    ) -> Result<Option<PrepareRenameResponse>> {
+        let symbol;
+        let doc_version;
+        {
+            let file = access.read(document);
+            doc_version = file.get_version();
+
+            let navigator = Navigator::new(&file);
+
+            symbol = navigator.get_symbol(pos);
+        }
+
+        if rename::renameable_symbol(&symbol) {
+            self.rename_cache
+                .set(document.to_owned(), doc_version, pos, symbol);
+            Ok(Some(PrepareRenameResponse::DefaultBehavior {
+                default_behavior: true,
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// `textDocument/rename` implementation.
+    ///
+    /// Renames symbol if possible. Checks that
+    /// there is no conflicts after rename.
+    ///
+    pub fn rename(
+        &self,
+        access: &WorkspaceAccess,
+        new_name: String,
+        pos: Position,
+        document: &Url,
+    ) -> Result<Option<WorkspaceEdit>> {
+        let file = access.read(document);
+        let navigator = Navigator::new(&file);
+
+        let symbol = self
+            .rename_cache
+            .get(document, file.get_version(), pos)
+            .map_or_else(|| navigator.get_symbol(pos), |cached| cached);
+
+        rename::renameable_to_symbol(&symbol, &new_name, file.get_elaborated())?;
+
+        let ranges = navigator.find_symbols(&symbol);
+
+        let edits = ranges
+            .into_iter()
+            .map(|s| {
+                Left(TextEdit {
+                    range: s,
+                    new_text: new_name.to_owned(),
+                })
+            })
+            .collect();
+
+        let text_document = OptionalVersionedTextDocumentIdentifier {
+            uri: document.to_owned(),
+            version: Some(file.get_version()),
+        };
+
+        let text_document_edits = TextDocumentEdit {
+            text_document,
+            edits,
+        };
+
+        let workspace_edit = WorkspaceEdit {
+            changes: None,
+            document_changes: Some(DocumentChanges::Edits(vec![text_document_edits])),
+            change_annotations: None,
+        };
+
+        Ok(Some(workspace_edit))
```

Comment:
I'd honestly just inline values which are used only once, like
```rust
Ok(Some(WorkspaceEdit {
    changes: None,
    change_annotations: None,
    document_changes: Some(DocumentChanges::Edits(vec![TextDocumentEdit {
        text_document: OptionalVersionedTextDocumentIdentifier {
            uri: document.to_owned(),
            version: Some(file.get_version()),
        },
        edits: ...,
    }])),
}))
```
So that the logic is not dispersed across the lines

Reply:
Don't know what to do with a formatter though...